### DISCUSSION TITLE
Bulk create for list of nested module

### DISCUSSION THREAD

**Discussion by avico78** [collerek/ormar#168] [Q&A]
trying to preform bulk_create for a nested module.
payment --> payment_means

By query below I'm getting list of payments from src DB
```python
    payments =  (await models_customer_fin_src.Payment.objects.select_all(follow=True)
                .filter(customer_no=customer_no).all()
            )
```

Not sure for how initialize list of payment by a module Payment on target DB so could preform .
Q: How can i initialize list with a Module?
bulk_create : 

```python
   payment_trg = models_customer_fin_trg.Payment(**payment)

   await payment_trg.bulk_create(follow=True, save_all=True)
```
Q - Can i preform bulk_create(follow=True, save_all=True) (same as done with save_all)for save the whole nested tree?

**Comment by collerek** [comment]
Unfortunately -> no, and its not `ormar` specific, but `sqlalchemy`/`databases` or rather even sql itself related.

When you choose bulk operation you sacrifice some functionality for speed of execution.

One of the functionalities you sacrifice is the return value. Meaning bulk operations (usually `execute_many`) does not return anything. That means that there is no way to fetch inserted `primary_key` values, and since you cannot fetch them you cannot save the whole tree, as relations would break. And it works the same in `sqlalchemy`, `databases` and any backend/db related lib I know, so it's not `ormar` specific.

In bulk_create in your example you would have to:
1.  save all payments in bulk
2.  update payments primary keys 
3.  bulk save all payments_means.

And since bulk operations don't update pks even if I would implement it (which would be quite complicated anyway given you need to save different models in different order depending on relation) the relations would brake (or even would raise exception in some cases).

So either you have to split it into two bulk create, when you load just inserted payments to retrieve their pks and only then add payment_means and bulk_create payment_means (but you would have to somehow assign proper means to payments).

If you would provide primary_keys for both models you can bulk_create them yet it's still needs to be divided in two calls - one per model.

Other way is unfortunately the loop and saving one by one.
